theano/gpuarray/opt.py

```diff
@@ -895,7 +903,34 @@ def local_gpua_subtensor(op, context_name, inputs, outputs):
                                               for v in n.inputs + n.outputs])
                         for n, _ in clients[0]]):
                     return
+    # Here is the condition for the GraphToGPU opt. inputs is the
+    # inputs we want to use for the new node
+    if (x.owner and isinstance(x.owner.op, GpuFromHost)):
+        cpu_x = x.owner.inputs[0]
+        # And it is a shared var or an input of the graph.
+        # and is used by only 1 node.
+        # x is in the new graph, so we can't tests its number of clients.
+        if not cpu_x.owner and len(cpu_x.clients) == 1:
+            return
+    return GpuSubtensor(op.idx_list)
+
 
+@register_opt2([tensor.Subtensor], 'fast_compile')
+def local_gpua_subtensor_graph(op, context_name, inputs, outputs):
+    # We need different code as the condition is different as inputs
+    # aren't the same.
```

Comment:
Those are special local optimizer. In fact, they aren't local optimizer, but function used by local optimier( the old opt) or by a global optimizer (the new opt). So we can define the interface we need. In the past, we used the same interface as local optimizer, but it was causing problem, so we created a new one.